### ISSUE TITLE
training code?

### ISSUE THREAD

**Issue by xinli2008** [alimama-creative/M3DDM-Video-Outpainting#1]
Hello, thank you for your nice job~! I am wondering will you release the training code?

**Comment by fanfanda** [comment]
Our training code is primarily borrowed from the training scripts found within the diffusers. Currently, we do not have plans to open-source our training code. Thank you for your interest and understanding.

**Comment by eshaanmoorjani** [comment]
The results from M3DDM are super cool :). Thanks for the work! 

Could you reference me to the training scripts that you modified for your implementation? I currently found a PyTorch implementation [here](https://github.com/lucidrains/make-a-video-pytorch/tree/main), but it doesn't use the diffusers library. It would be nice to use diffusers

**Comment by culeao** [comment]
> The results from M3DDM are super cool :). Thanks for the work!
> 
> Could you reference me to the training scripts that you modified for your implementation? I currently found a PyTorch implementation [here](https://github.com/lucidrains/make-a-video-pytorch/tree/main), but it doesn't use the diffusers library. It would be nice to use diffusers

Thank you for your attention!

Sorry to say that we don't have plans to release a training script, but in fact, the training script is very simple, just replace our model based on the training examples in diffusers (e.g. [train_text_to_image](https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image.py)). You need to do processing on the training video data, e.g., processing to get _mask_image_ and _mask_.

**Comment by eshaanmoorjani** [comment]
Thanks! We rewrote the training code but for some reason, our outpaint is having a line when it outpaints downwards. Any ideas why here? 
<img width="481" alt="Screenshot 2024-02-20 at 2 02 55 PM" src="https://github.com/alimama-creative/M3DDM-Video-Outpainting/assets/71729984/f4db170d-c697-4fca-87ca-b53d0005faa6">

**Comment by fanfanda** [comment]
We have not encountered this situation before. I recommend providing more details on training and inference parameters. 

If you are using our inference pipeline, you can try changing the use_add_noise parameter to False in `outpainting_with_random_masked_latent_inference_bidirection` function.

**Comment by fanfanda** [comment]
> Thanks! We rewrote the training code but for some reason, our outpaint is having a line when it outpaints downwards. Any ideas why here? <img alt="Screenshot 2024-02-20 at 2 02 55 PM" width="481" src="https://private-user-images.githubusercontent.com/71729984/306429117-f4db170d-c697-4fca-87ca-b53d0005faa6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg0ODE0MTYsIm5iZiI6MTcwODQ4MTExNiwicGF0aCI6Ii83MTcyOTk4NC8zMDY0MjkxMTctZjRkYjE3MGQtYzY5Ny00ZmNhLTg3Y2EtYjUzZDAwMDVmYWE2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjIxVDAyMDUxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZiNzY2NWViNjgzNzhlNTU2NTBhN2JiZTE0ZjQ4M2ZlODliZThhZjM0ZDljYzI2MmFlNjllOTNmODc1ZTI3MmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.vgS4CBxQ_7mlGngk5NIsEV6DQ1bvMqdVzGrn6fVDaqk">

We have not encountered this situation before. We recommend providing more details on training and inference parameters.

If you are using our inference pipeline, you can try changing the use_add_noise parameter to False in outpainting_with_random_masked_latent_inference_bidirection function.

**Comment by jimmyl02** [comment]
following up here's our exact training setup:
- we have a full image (4 dims), a masked image where the masked regions are all 0s (4 dims), and a mask where masked region is 1s and unmasked is 0s (1 dim)
- we add noise according to the scheduler to the first 4 dimensions with a random timestep embedding
- our loss objective is the noise added 
- we have 25% chance of it being fully masked and 10% chance of fully masked global frames for CFG
- we follow the masking distribution in the paper

any advice on steps that are wrong / could be improved would be extremely helpful, thanks!

**Comment by fanfanda** [comment]
> following up here's our exact training setup:
> 
> * we have a full image (4 dims), a masked image where the masked regions are all 0s (4 dims), and a mask where masked region is 1s and unmasked is 0s (1 dim)
> * we add noise according to the scheduler to the first 4 dimensions with a random timestep embedding
> * our loss objective is the noise added
> * we have 25% chance of it being fully masked and 10% chance of fully masked global frames for CFG
> * we follow the masking distribution in the paper
> 
> any advice on steps that are wrong / could be improved would be extremely helpful, thanks!

It looks fine.

**Comment by eshaanmoorjani** [comment]
Another question -- how did y'all achieve a batch size of 10 per GPU? (I'm assuming batch size of 240 / 24 A100s = a batch size of 10 per GPU). Right now, we're returning 3 tensors (labels, masked frames, and masks).

**Comment by fanfanda** [comment]
> Another question -- how did y'all achieve a batch size of 10 per GPU? (I'm assuming batch size of 240 / 24 A100s = a batch size of 10 per GPU). Right now, we're returning 3 tensors (labels, masked frames, and masks).

Yes, the batch size on each card is 10. Correspondingly, the batch dimension of each tensor is also 10. 
We have an engineering team that has developed a very good distributed training system, so we haven't had to focus on issues related to distributed training.
